okapi-core/src/main/java/org/folio/okapi/managers/ProxyService.java

```diff
@@ -592,7 +592,8 @@ private void proxyResponseImmediate(ProxyContext pc, ReadStream<Buffer> res,
         for (HttpClientRequest r : clientRequestList) {
           r.write(data);
         }
-        ctx.response().write(data);
+        res.pause();
+        ctx.response().write(data, end -> res.resume());
```

Comment:
write(Buffer, Handler) does NOT wait if the queue is full. It calls the Handler after appending Buffer to the queue regardless of the queue status.
The documenation
* https://vertx.io/docs/apidocs/io/vertx/core/streams/WriteStream.html
* https://vertx.io/docs/vertx-core/java/#streams

explains to call writeQueueFull() to get a signal about a full queue.

You probably need to implement
* https://vertx.io/docs/apidocs/io/vertx/core/streams/Pump.html
* https://github.com/eclipse-vertx/vert.x/blob/master/src/main/java/io/vertx/core/streams/impl/PipeImpl.java

with one ReadStream and multiple WriteStreams.

Reply:
Yes. That's why it has to pile up in memory if it can't immediately write it.

Reply:
write(Buffer data, Handler<AsyncResult<Void>> handler)

Same as write(Buffer) but with an handler called when the operation completes

*and* and FWIW, Hongwei sees no memory issues.

Reply:
I can try to make a solution that does not use write with handler, but with writeQueueFull and drainHandler.

Reply:
OK. Giving up on that. drainHandler is a NetSocket only thing.

Reply:
OK.. Now in place.. More complicated than before.. And as far as I can see results are similar, memory wise. Hongwei, you'll have to test this one.

Reply:
Tested. Still works.

Reply:
```
package tmp;

import org.apache.commons.lang3.StringUtils;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import io.vertx.core.Vertx;
import io.vertx.core.http.HttpClientRequest;
import io.vertx.junit5.VertxExtension;
import io.vertx.junit5.VertxTestContext;

@ExtendWith(VertxExtension.class)
class StreamTest {
  String chunk = StringUtils.repeat("chunk ", 20000);

  void writeChunk(int i, HttpClientRequest request) {
    request.write(chunk, res -> writeChunk(i+1, request));
    System.out.println(i + " " + request.writeQueueFull() + " ");
  }

  @Test()
  void test(Vertx vertx, VertxTestContext vtc) {
    vertx
    .createHttpServer()
    .requestHandler(req -> {
      req.pause();
      // don't fetch the data
    })
    .listen(8888, server -> {
      HttpClientRequest request = vertx
      .createHttpClient()
      .post(8888, "localhost", "/")
      .setChunked(true)
      .handler(handler -> {});
      writeChunk(1, request);
    });
  }
}
package tmp;

import org.apache.commons.lang3.StringUtils;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import io.vertx.core.Vertx;
import io.vertx.core.http.HttpClientRequest;
import io.vertx.junit5.VertxExtension;
import io.vertx.junit5.VertxTestContext;

@ExtendWith(VertxExtension.class)
class StreamTest {
  String chunk = StringUtils.repeat("chunk ", 20000);

  void writeChunk(int i, HttpClientRequest request) {
    request.write(chunk, res -> writeChunk(i+1, request));
    System.out.println(i + " " + request.writeQueueFull() + " ");
  }

  @Test()
  void test(Vertx vertx, VertxTestContext vtc) {
    vertx
    .createHttpServer()
    .requestHandler(req -> {
      req.pause();
      // don't fetch the data
    })
    .listen(8888, server -> {
      HttpClientRequest request = vertx
      .createHttpClient()
      .post(8888, "localhost", "/")
      .setChunked(true)
      .handler(handler -> {});
      writeChunk(1, request);
    });
  }
}
```
Output:
```
1 false 
2 true 
3 true 
4 true 
...
```